proposals/text/3784-input-output-shaper.md

```diff
@@ -0,0 +1,247 @@
+These new variables hold values which are a result of some arbitrary function invocation on the existing variables
+in the invocation context.
+
+InputOutputShaper is especially useful when combined with PromptNodes. Aside from simply renaming variables to match
+the templates of PromptNodes, we can also use InputOutputShaper to add new variables to the invocation context. Often
+these new variables are the result of some arbitrary function invocation on the existing variables in the invocation.
+
+The original idea for InputOutputShaper was to support the use case of expanding the query string variable to a list of
+strings and the list size matching the size of the documents list. This is useful when we want to use the same query string
+as a question to pose to all the documents in the documents list.
+
+The expected outcome of using InputOutputShaper is that we can easily add new variables to the invocation context
+so they can match the prompt templates of PromptNodes. Multiple InputOutputShaper components can be used in a pipeline to
+modify the invocation context as needed.
+
+
+# Detailed design
+
+The InputOutputShaper component is primarily used in pipelines and configured via a YAML file. The YAML component
+definition consists of the params block:
+
+```yaml
+            components:
+            - name: shaper
+              params:
+                inputs:
+                   query:
+                      func: expand
+                      output: questions
+                      params:
+                        expand_target: query
+                        size:
+                          func: len
+                          params:
+                            - documents
+                   documents:
+                      func: contract_docs
+                      params:
+                        docs: documents
+                        delimiter: " "
+              type: InputOutputShaper
+            pipelines:
+              - name: query
+                nodes:
+                  - name: shaper
+                    inputs:
+                      - Query
+```
+
+The params block consists of the inputs list. Each item in the inputs list is a dictionary with the key being the
+invocation context variable that we want to modify.
+
+In the example above, we have two items in the inputs list. The first item is a dictionary with the key `query` and the
+second item is a dictionary with the key `documents`.
+
+For the query variable, we want to invoke the function `expand` and store the result in the variable `questions`.
+The `expand` function takes two keyword parameters: `expand_target` and `size`. The `expand_target` parameter is the
+name of the variable in the invocation context that we want to expand. The `size`parameter is a result of the `len`
+function invocation on the variable `documents`.
+
+For the documents variable, we want to invoke the function `contract_docs` and store the result in the same variable.
```

Comment:
I find the naming `contract_docs` slightly confusing, too (thinking of contracts).

Reply:
Yeah, Julian was spot on as [well](https://github.com/deepset-ai/haystack/pull/3784#discussion_r1059899887)